conda_libmamba_solver/solver.py

```diff
@@ -475,21 +478,22 @@ def _specs_to_tasks_add(self, in_state: SolverInputState, out_state: SolverOutpu
                 tasks[("USERINSTALLED", api.SOLVER_USERINSTALLED)].append(installed_spec_str)
 
             # These specs are explicit in some sort of way
-            if pinned:
+            if pinned and not pinned.is_name_only_spec:
                 # these are the EXPLICIT pins; conda also uses implicit pinning to
                 # constrain updates too but those can be overridden in case of conflicts.
-                if pinned.is_name_only_spec:
-                    # pins need to constrain in some way, otherwide is undefined behaviour
-                    pass
-                elif requested and not requested.match(pinned):
```

Comment:
This is the faulty line that triggered the whole investigation. `MatchSpec.match(MatchSpec)` seems to work in simple cases but that's mostly accidental. It's not the intended usage (it should do `MatchSpec.match(PackageRecord)`). 

This line _could_ have been replaced with something like:

```python
elif requested and not compatible_matchspecs(requested, pinned, index):
```

where `compatible_matchspecs()` would query the `index` with the two specs and compute the intersection of the returned `PackageRecord` lists. If not null, then both specs are deemed compatible.

However, we decided against this due to the RAM overheads of the querying (`index.query("python")` is a HUGE JSON payload). It would have been okayish if there was an iterator interface with one package at a time, but well, that's what it is.

Adding this comment for posterity and an explanation of the choices we did here.

Reply:
Update: we ended up doing `compatible_specs`. It's not that bad of an overhead (see #294).